src/main/scala/analysis/IDEAnalysis.scala

```diff
@@ -0,0 +1,61 @@
+package analysis
+
+import ir.{CFGPosition, Command, DirectCall, GoTo, Procedure, Program}
+
+final case class Lambda()
+
+/**
+ * Base trait for IDE analyses.
+ *
+ * @tparam E Type of the function entry CFGPosition
+ * @tparam EE Type of the function exit CFGPosition
+ * @tparam C Type of a function call
+ * @tparam R Type of a call return site
+ * @tparam D the type of items
+ * @tparam T type of elements of the value lattice
+ * @tparam L the type of the value lattice
+ * Adapted from Tip
+ * https://github.com/cs-au-dk/TIP/blob/master/src/tip/solvers/IDEAnalysis.scala
+ * The special item representing the empty element in IDE.
+ */
+trait IDEAnalysis[E, EE, C, R, D, T, L <: Lattice[T]] {
+  val program: Program
+
+  type DL = Either[D, Lambda]
+
+  /**
+   * The value lattice.
+   */
+  val valuelattice: L
+
+  /**
+   * The edge lattice.
+   */
+  val edgelattice: EdgeFunctionLattice[T, valuelattice.type]
+
+  /**
+   * Edges for call-to-entry.
+   */
+  def edgesCallToEntry(call: C, entry: E)(d: DL): Map[DL, edgelattice.Element]
+
+  /**
+   * Edges for exit-to-aftercall.
+   */
+  def edgesExitToAfterCall(exit: EE, aftercall: R)(d: DL): Map[DL, edgelattice.Element]
+
+  /**
+   * Edges for call-to-aftercall.
+   */
+  def edgesCallToAfterCall(call: C, aftercall: R)(d: DL): Map[DL, edgelattice.Element]
+
+  /**
+   * Edges for other CFG nodes.
+   */
+  def edgesOther(n: CFGPosition)(d: DL): Map[DL, edgelattice.Element]
+}
+
+
+trait ForwardIDEAnalysis[D, T, L <: Lattice[T]] extends IDEAnalysis[Procedure, Command, DirectCall, GoTo, D, T, L]
+
+
+trait BackwardIDEAnalysis[D, T, L <: Lattice[T]] extends IDEAnalysis[Command, Procedure, GoTo, DirectCall, D, T, L]
```

Comment:
Shouldn't we be able to have stricter type constraints on these? The exit for forward analysis/entrance for backward analysis should be a return shouldn't it? (which is for now an IndirectCall)

@sadrabt

Reply:
Most likely. I never came across examples where the exit for forward analysis/entrance for backward analysis is not a return/IndirectCall from our examples.

this is not the case in stripped/uncalled functions so I'm not sure if there is a program that will have something else there
![Screenshot from 2024-02-29 13-51-39](https://github.com/UQ-PAC/bil-to-boogie-translator/assets/69196526/d8cda522-7656-4719-a6a6-a70d6deadf19)

Either way, the analysis won't cope well if a function doesn't have a returnblock like above, so the type should be tighter 
(indirectCall/ new return node)